src/crimson/os/seastore/linked_tree_node.h

```diff
@@ -0,0 +1,973 @@
+	    t, dynamic_cast<CachedExtent*>(child));
+	}
+      } else {
+	return true;
+      }
+    } else {
+      return true;
+    }
+  }
+
+  parent_tracker_t<T>* my_tracker = nullptr;
+private:
+  T& down_cast() {
+    return *static_cast<T*>(this);
+  }
+  const T& down_cast() const {
+    return *static_cast<const T*>(this);
+  }
+
+  std::vector<BaseChildNode<T, node_key_t>*> children;
+  std::set<TCachedExtentRef<T>, Comparator> copy_sources;
+  uint16_t capacity = 0;
+
+  // copy dests points from a stable node back to its pending nodes
+  // having copy sources at the same tree level, it serves as a two-level index:
+  // transaction-id then node-key to the pending node.
+  //
+  // The copy dest pointers must be symmetric to the copy source pointers.
+  //
+  // copy_dests_t will be automatically unregisterred upon transaction destruction,
+  // see Transaction::views
+  struct copy_dests_t : trans_spec_view_t {
+    std::set<TCachedExtentRef<T>, Comparator> dests_by_key;
+    copy_dests_t(Transaction &t) : trans_spec_view_t{t.get_trans_id()} {}
+    ~copy_dests_t() {
+      LOG_PREFIX(~copy_dests_t);
+      SUBTRACE(seastore_fixedkv_tree, "copy_dests_t destroyed");
+    }
+  };
+
+  using trans_view_set_t = trans_spec_view_t::trans_view_set_t;
+  trans_view_set_t copy_dests_by_trans;
+  CacheProxyRef cache;
+  template <typename, typename, typename>
+  friend class ChildNode;
+  template <typename>
+  friend class parent_tracker_t;
+  template <typename>
+  friend class child_pos_t;
+#ifndef UNIT_TESTS_BUILT
+  template <typename, typename, typename, typename, typename, size_t, bool>
+  friend class FixedKVBtree;
+#endif
+};
+
+// ChildNodes are non-root nodes in the tree or extents that
+// have parents, e.g. LogicalCachedExtents have LBALeafnodes
+// as the parents, so they are ChildNodes.
+template <typename ParentT, typename ChildT, typename key_t>
+class ChildNode : public BaseChildNode<ParentT, key_t> {
+protected:
+  void take_prior_parent_tracker() {
+    auto &me = down_cast();
+    auto &prior = static_cast<ChildT&>(*me.get_prior_instance());
+    this->parent_tracker = prior.parent_tracker;
+  }
+
+  void set_parent_tracker_from_prior_instance() {
+    auto &me = down_cast();
+    assert(!me.is_root());
+    assert(me.is_mutation_pending());
+    take_prior_parent_tracker();
+    assert(me.is_parent_valid());
+    auto parent = this->get_parent_node();
+    //TODO: can this search be avoided?
+    auto off = parent->lower_bound(this->get_begin()).get_offset();
+    assert(parent->iter_idx(off).get_key() == this->get_begin());
+    parent->children[off] = &me;
+  }
+
+  void on_initial_write() {
+    auto &me = down_cast();
+    if (me.is_root()) {
+      this->reset_parent_tracker();
+    }
+  }
+  void on_invalidated() {
+    this->reset_parent_tracker();
+  }
+  void on_replace_prior() {
+    set_parent_tracker_from_prior_instance();
+  }
+  void destroy() {
+    auto &me = down_cast();
+    if (!me.is_root()) {
```

Comment:
Seems to be another example due to `ChildNode` and `RootNode` are not properly abstracted together.

Reply:
I think this should be changed to an assertion, the judgement should be outside linked tree node, what do you think?